### DISCUSSION TITLE
Canarytrace pri dobehnuti neuzavre Selenium container

### DISCUSSION THREAD

**Discussion by marekvesely-direct** [canarytrace/documentation#194] [Synthetic]
Dobry den,

canarytrace pri dobehnuti jobu necha otevreny container selenium a cely pod zustane ve stavu not ready. Ve vysledku to znamena ze HW zdroje jsou pro tento container stale rezervovane i kdyz nejsou potreba, coz je problem. Spravne by po dobehu se mel cely pod uzavrit a oznacit jako Completed. Poprosim o napravu at zbytecne neplitvame drahym zelezem v cloudu.

Dekuji,

Marek

**Comment by rdpanek** [comment]
Dobrý den, 
oba kontejnery jsou bez stavové a tím pádem o sobě nijak nevědí a druhý kontejner s prohlížečem nemůže vědět, že váš test nebude pokračovat druhou instancí.

První možností jak dříve ukončit browser je zkrátit čas v CronJobu. Pokud potřebujete vaše flow pozměnit, můžete využít Kubernetes API.

Pro tento účel může kontejner, který má ukončit druhý kontejner, využít Kubernetes API k odeslání příkazu k ukončení. Zde je několik kroků, jak toho můžete dosáhnout:

### Krok 1: Připravte Kubernetes API Klienta

Váš kontejner bude potřebovat klienta pro komunikaci s Kubernetes API. To může být klientská knihovna v jazyce, který používáte, nebo můžete použít `kubectl` přímo z kontejneru (i když to není obvykle doporučováno z bezpečnostních důvodů).

### Krok 2: Přidělení Přístupových Práv

Váš kontejner bude potřebovat přístupová práva k ukončení jiného kontejneru. To lze dosáhnout vytvořením role a role bindingu, který dává kontejneru potřebná oprávnění.

### Krok 3: Ukonečení Kontejneru

Z kontejneru, který má ukončit druhý kontejner, můžete vytvořit požadavek na Kubernetes API k ukončení kontejneru. To může vypadat nějak takto (předpokládá se, že máte k dispozici `kubectl` a potřebná oprávnění):

```bash
kubectl exec <POD_NAME> -c <CONTAINER_NAME_TO_BE_TERMINATED> -- /bin/sh -c "pkill -TERM -f <PROCESS_NAME_IN_CONTAINER>"
```

Alternativně, můžete použít programovací knihovnu ke komunikaci s API, zde je příklad v Pythonu s použitím kubernetes klienta:

```python
from kubernetes import client, config

# Nainicializujte klienta
config.load_kube_config()
v1 = client.CoreV1Api()

# Ukonečte proces v kontejneru
exec_response = v1.connect_post_namespaced_pod_exec(
    name='POD_NAME',
    namespace='NAMESPACE',
    command=['/bin/sh', '-c', 'pkill -TERM -f PROCESS_NAME_IN_CONTAINER'],
    container='CONTAINER_NAME_TO_BE_TERMINATED',
    stderr=True, stdin=False,
    stdout=True, tty=False
)

print(exec_response)
```

Prosím, všimněte si, že `<PROCESS_NAME_IN_CONTAINER>` by měl být název procesu, který chcete ukončit v cílovém kontejneru.
